File: win-prog/ProgWinVB/Toolbars and Status Bars/MenuHelpFirstTry/MenuHelpFirstTry.vb
'-------------------------------------------------
' MenuHelpFirstTry.vb (c) 2002 by Charles Petzold
'-------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class MenuHelpFirstTry
    Inherits Form

    Private sbpMenuHelp As StatusBarPanel
    Private strSavePanelText As String

    Shared Sub Main()
        Application.Run(New MenuHelpFirstTry())
    End Sub

    Sub New()
        Text = "Menu Help (First Try)"
        BackColor = SystemColors.Window
        ForeColor = SystemColors.WindowText

        ' Create a status bar with one panel.
        Dim sbar As New StatusBar()
        sbar.Parent = Me
        sbar.ShowPanels = True
        sbpMenuHelp = New StatusBarPanel()
        sbpMenuHelp.Text = "Ready"
        sbpMenuHelp.AutoSize = StatusBarPanelAutoSize.Spring
        sbar.Panels.Add(sbpMenuHelp)

        ' Construct a simple menu.
        ' For this demo, we can ignore the Click handlers
        '   but what we really need are Select handlers.
        Menu = New MainMenu()
        Dim ehSelect As EventHandler = AddressOf MenuOnSelect

        ' Create File menu items.
        Dim mi As New MenuItem("File")
        AddHandler mi.Select, ehSelect
        Menu.MenuItems.Add(mi)

        mi = New MenuItem("Open")
        AddHandler mi.Select, ehSelect
        Menu.MenuItems(0).MenuItems.Add(mi)

        mi = New MenuItem("Close")
        AddHandler mi.Select, ehSelect
        Menu.MenuItems(0).MenuItems.Add(mi)

        mi = New MenuItem("Save")
        AddHandler mi.Select, ehSelect
        Menu.MenuItems(0).MenuItems.Add(mi)

        ' Create Edit menu items.
        mi = New MenuItem("Edit")
        AddHandler mi.Select, ehSelect
        Menu.MenuItems.Add(mi)

        mi = New MenuItem("Cut")
        AddHandler mi.Select, ehSelect
        Menu.MenuItems(1).MenuItems.Add(mi)

        mi = New MenuItem("Copy")
        AddHandler mi.Select, ehSelect
        Menu.MenuItems(1).MenuItems.Add(mi)

        mi = New MenuItem("Paste")
        AddHandler mi.Select, ehSelect
        Menu.MenuItems(1).MenuItems.Add(mi)
    End Sub

    Protected Overrides Sub OnMenuStart(ByVal ea As EventArgs)
        strSavePanelText = sbpMenuHelp.Text
    End Sub

    Protected Overrides Sub OnMenuComplete(ByVal ea As EventArgs)
        sbpMenuHelp.Text = strSavePanelText
    End Sub

    Private Sub MenuOnSelect(ByVal obj As Object, ByVal ea As EventArgs)
        Dim mi As MenuItem = DirectCast(obj, MenuItem)
        Dim str As String

        Select Case mi.Text
            Case "File" : str = "Commands for working with files"
            Case "Open" : str = "Opens an existing document"
            Case "Close" : str = "Closes the current document"
            Case "Save" : str = "Saves the current document"
            Case "Edit" : str = "Commands for editing the document"
            Case "Cut" : str = "Deletes the selection and " & _
                               "copies it to the clipboard"
            Case "Copy" : str = "Copies the selection to the " & _
                                "clipboard"
            Case "Paste" : str = "Replaces the current selection " & _
                                 "with the clipboard contents"
            Case Else : str = ""
        End Select

        sbpMenuHelp.Text = str
    End Sub
End Class

File: win-prog/ProgWinCS/Toolbars and Status Bars/MenuHelpFirstTry/MenuHelpFirstTry.cs
//-----------------------------------------------
// MenuHelpFirstTry.cs © 2001 by Charles Petzold
//-----------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class MenuHelpFirstTry: Form
{
     StatusBarPanel sbpMenuHelp;
     string         strSavePanelText;

     public static void Main()
     {
          Application.Run(new MenuHelpFirstTry());
     }
     public MenuHelpFirstTry()
     {
          Text = "Menu Help (First Try)";
          BackColor = SystemColors.Window;
          ForeColor = SystemColors.WindowText;

               // Create a status bar with one panel.

          StatusBar sb = new StatusBar();
          sb.Parent = this;
          sb.ShowPanels = true;

          sbpMenuHelp = new StatusBarPanel();
          sbpMenuHelp.Text = "Ready";
          sbpMenuHelp.AutoSize = StatusBarPanelAutoSize.Spring;

          sb.Panels.Add(sbpMenuHelp);

               // Construct a simple menu.
               // For this demo, we can ignore the Click handlers
               //   but what we really need are Select handlers.

          Menu = new MainMenu();
          EventHandler ehSelect = new EventHandler(MenuOnSelect);
          
               // Create File menu items.

          MenuItem mi = new MenuItem("File");
          mi.Select += ehSelect;
          Menu.MenuItems.Add(mi);

          mi = new MenuItem("Open");
          mi.Select += ehSelect;
          Menu.MenuItems[0].MenuItems.Add(mi);
          
          mi = new MenuItem("Close");
          mi.Select += ehSelect;
          Menu.MenuItems[0].MenuItems.Add(mi);

          mi = new MenuItem("Save");
          mi.Select += ehSelect;
          Menu.MenuItems[0].MenuItems.Add(mi);

               // Create Edit menu items.

          mi = new MenuItem("Edit");
          mi.Select += ehSelect;
          Menu.MenuItems.Add(mi);

          mi = new MenuItem("Cut");
          mi.Select += ehSelect;
          Menu.MenuItems[1].MenuItems.Add(mi);
          
          mi = new MenuItem("Copy");
          mi.Select += ehSelect;
          Menu.MenuItems[1].MenuItems.Add(mi);

          mi = new MenuItem("Paste");
          mi.Select += ehSelect;
          Menu.MenuItems[1].MenuItems.Add(mi);
     }
     protected override void OnMenuStart(EventArgs ea)
     {
          strSavePanelText = sbpMenuHelp.Text;
     }
     protected override void OnMenuComplete(EventArgs ea)
     {
          sbpMenuHelp.Text = strSavePanelText;
     }
     void MenuOnSelect(object obj, EventArgs ea)
     {
          MenuItem mi = (MenuItem) obj;
          string   str;

          switch (mi.Text)
          {
          case "File":   str = "Commands for working with files";    break;
          case "Open":   str = "Opens an existing document";         break;
          case "Close":  str = "Closes the current document";        break;
          case "Save":   str = "Saves the current document";         break;
          case "Edit":   str = "Commands for editing the document";  break;
          case "Cut":    str = "Deletes the selection and " +
                               "copies it to the clipboard";         break;
          case "Copy":   str = "Copies the selection to the " +
                               "clipboard";                          break;
          case "Paste":  str = "Replaces the current selection " +
                               "with the clipboard contents";        break;
          default:       str = "";                                   break;
          }

          sbpMenuHelp.Text = str;
     }
}
